File: app/src/main/java/com/icerrate/popularmovies/view/movies/search/SearchMoviesActivity.java
package com.icerrate.popularmovies.view.movies.search;

import android.content.Context;
import android.content.Intent;
import android.content.pm.ActivityInfo;
import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.support.v4.app.ActivityOptionsCompat;
import android.support.v4.view.MenuItemCompat;
import android.support.v4.view.ViewCompat;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SearchView;
import android.view.Display;
import android.view.Menu;
import android.view.MenuItem;
import android.view.Surface;
import android.view.View;
import android.view.ViewStub;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.icerrate.popularmovies.R;
import com.icerrate.popularmovies.data.model.Movie;
import com.icerrate.popularmovies.data.model.PaginatedResponse;
import com.icerrate.popularmovies.utils.ViewUtils;
import com.icerrate.popularmovies.utils.InjectionUtils;
import com.icerrate.popularmovies.view.common.BaseActivity;
import com.icerrate.popularmovies.view.common.EndlessRecyclerOnScrollListener;
import com.icerrate.popularmovies.view.movies.catalog.MoviesCatalogAdapter;
import com.icerrate.popularmovies.view.movies.detail.MovieDetailActivity;

import java.util.List;
import java.util.Objects;

import butterknife.BindView;

import static com.icerrate.popularmovies.view.movies.detail.MovieDetailFragment.KEY_MOVIE;

/**
 * @author Ivan Cerrate.
 */

public class SearchMoviesActivity extends BaseActivity implements SearchMoviesContract.View,
        MoviesCatalogAdapter.OnItemClickListener {

    public final static String KEY_ACTIVE_SEARCH = "ACTIVE_SEARCH_KEY";

    public final static String KEY_SEARCH_QUERY = "SEARCH_QUERY_KEY";

    public final static String KEY_PAGINATED_MOVIES = "PAGINATED_MOVIES_KEY";

    @BindView(R.id.search_view)
    public SearchView searchView;

    @BindView(R.id.progress)
    public ViewStub progressBar;

    @BindView(R.id.movies)
    public RecyclerView moviesRecyclerView;

    @BindView(R.id.movies_no_data)
    public TextView noDataTextView;

    @BindView(R.id.footer_progress)
    public ViewStub footerProgressBar;

    private ImageView selectedMoviePoster;

    private MoviesCatalogAdapter adapter;

    private EndlessRecyclerOnScrollListener endlessRecyclerOnScrollListener;

    private SearchMoviesPresenter presenter;

    public static Intent makeIntent(Context context){
        return new Intent(context, SearchMoviesActivity.class);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_search_movies);
        setNavigationToolbar(true);
        presenter = new SearchMoviesPresenter(this,
                InjectionUtils.movieRepository(this));
        setupView();
        if (savedInstanceState != null) {
            restoreInstanceState(savedInstanceState);
            presenter.loadMovies();
            searchView.post(new Runnable() {
                @Override
                public void run() {
                    searchView.setQuery(presenter.getQuery(), false);
                }
            });
        }
    }

    private void setupView() {
        //Movies
        int columns = ViewUtils.getGridColumns(this);
        GridLayoutManager gridLayoutManager = new GridLayoutManager(this, columns);
        endlessRecyclerOnScrollListener = new EndlessRecyclerOnScrollListener(gridLayoutManager) {
            @Override
            public void onLoadMore() {
                presenter.loadNextMoviesPage();
            }
        };
        adapter = new MoviesCatalogAdapter(this, columns);
        moviesRecyclerView.setAdapter(adapter);
        moviesRecyclerView.setLayoutManager(gridLayoutManager);
        moviesRecyclerView.addOnScrollListener(endlessRecyclerOnScrollListener);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        //SearchView
        searchView.onActionViewExpanded();
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                return false;
            }
            @Override
            public boolean onQueryTextChange(String newText) {
                presenter.searchMovies(newText);
                return false;
            }
        });
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    protected void saveInstanceState(Bundle outState) {
        outState.putString(KEY_SEARCH_QUERY, presenter.getQuery());
        outState.putParcelable(KEY_PAGINATED_MOVIES, presenter.getMoviesPaginatedResponse());
    }

    @Override
    protected void restoreInstanceState(Bundle savedState) {
        String query = savedState.getString(KEY_SEARCH_QUERY);
        PaginatedResponse<Movie> moviesPaginatedResponse = savedState.getParcelable(KEY_PAGINATED_MOVIES);
        presenter.loadPresenterState(query, moviesPaginatedResponse);
    }

    @Override
    public void showProgressBar(boolean show) {
        if (progressBar != null) {
            progressBar.setVisibility(show ? View.VISIBLE : View.GONE);
        }
    }

    @Override
    public void showError(String errorMessage) {
        ViewUtils.createSnackbar(toolbar, errorMessage, Snackbar.LENGTH_SHORT).show();
    }

    @Override
    public void resetMovies() {
        adapter.resetItems();
        adapter.setLoading(false);
        endlessRecyclerOnScrollListener.setLoading(false);
    }

    @Override
    public void showMovies(List<Movie> movies) {
        adapter.addItems(movies);
        endlessRecyclerOnScrollListener.setLoading(false);
    }

    @Override
    public void showSearchHint(boolean show) {
        noDataTextView.setVisibility(show ? View.VISIBLE : View.GONE);
        noDataTextView.setText(getString(R.string.search_movies_hint));
    }

    @Override
    public void showNoDataView(boolean show) {
        noDataTextView.setVisibility(show ? View.VISIBLE : View.GONE);
        noDataTextView.setText(getString(R.string.movies_no_data));
    }

    @Override
    public void showFooterProgress(boolean show) {
        adapter.setLoading(show);
        footerProgressBar.setVisibility(show ? View.VISIBLE : View.GONE);
    }

    @Override
    public void goToMovieDetail(Movie movie) {
        ActivityOptionsCompat options = ActivityOptionsCompat.makeSceneTransitionAnimation(this,
                selectedMoviePoster, Objects.requireNonNull(ViewCompat.getTransitionName(selectedMoviePoster)));
        startActivity(MovieDetailActivity.makeIntent(this)
                .putExtra(KEY_MOVIE, movie), options.toBundle());
        overridePendingTransition(R.anim.fade_in, R.anim.none);
    }

    @Override
    public void onItemClick(View view) {
        Movie movie = (Movie) view.getTag();
        if (movie != null) {
            selectedMoviePoster = view.findViewById(R.id.poster);
            presenter.onMovieItemClick(movie);
        }
    }
}

File: app/src/main/java/com/icerrate/popularmovies/view/movies/search/SearchMoviesActivity.kt
package com.icerrate.popularmovies.view.movies.search

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import androidx.activity.viewModels
import androidx.core.view.MenuProvider
import androidx.lifecycle.Lifecycle
import dagger.hilt.android.AndroidEntryPoint
import android.widget.ImageView
import androidx.appcompat.widget.SearchView
import androidx.core.app.ActivityOptionsCompat
import androidx.core.view.ViewCompat
import androidx.recyclerview.widget.GridLayoutManager
import com.google.android.material.snackbar.Snackbar
import com.icerrate.popularmovies.databinding.ActivitySearchMoviesBinding
import com.icerrate.popularmovies.R
import com.icerrate.popularmovies.data.model.Movie
import com.icerrate.popularmovies.data.model.PaginatedResponse
import com.icerrate.popularmovies.utils.ViewUtils
import com.icerrate.popularmovies.view.common.BaseActivity
import com.icerrate.popularmovies.view.common.EndlessRecyclerOnScrollListener
import com.icerrate.popularmovies.view.movies.catalog.MoviesCatalogAdapter
import com.icerrate.popularmovies.view.movies.detail.MovieDetailActivity
import com.icerrate.popularmovies.view.movies.detail.MovieDetailFragment.Companion.KEY_MOVIE

/**
 * @author Ivan Cerrate.
 */
@AndroidEntryPoint
class SearchMoviesActivity : BaseActivity<ActivitySearchMoviesBinding>(
    ActivitySearchMoviesBinding::inflate
), MoviesCatalogAdapter.OnItemClickListener, MenuProvider {

    private var selectedMoviePoster: ImageView? = null
    private lateinit var adapter: MoviesCatalogAdapter
    private lateinit var endlessRecyclerOnScrollListener: EndlessRecyclerOnScrollListener
    private val viewModel: SearchMoviesViewModel by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        addMenuProvider(this, this, Lifecycle.State.RESUMED)

        setNavigationToolbar(true)
        setupView()
        setupSearchView()
        observeViewModel()
        
        if (savedInstanceState != null) {
            restoreInstanceState(savedInstanceState)
            viewModel.processIntent(SearchMoviesIntent.LoadMovies)
            binding.searchView.post {
                binding.searchView.setQuery(viewModel.state.value?.query ?: "", false)
            }
        }
    }

    private fun setupView() {
        // Movies
        val columns = ViewUtils.getGridColumns(this)
        val gridLayoutManager = GridLayoutManager(this, columns)
        endlessRecyclerOnScrollListener = object : EndlessRecyclerOnScrollListener(gridLayoutManager) {
            override fun onLoadMore() {
                viewModel.processIntent(SearchMoviesIntent.LoadNextMoviesPage)
            }
        }
        adapter = MoviesCatalogAdapter(onItemClickListener = this, columns = columns)
        binding.movies.adapter = adapter
        binding.movies.layoutManager = gridLayoutManager
        binding.movies.addOnScrollListener(endlessRecyclerOnScrollListener)
    }

    private fun setupSearchView() {
        binding.searchView.onActionViewExpanded()
        binding.searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean = false

            override fun onQueryTextChange(newText: String): Boolean {
                viewModel.processIntent(SearchMoviesIntent.SearchMovies(newText))
                return false
            }
        })
    }

    private fun observeViewModel() {
        viewModel.state.observe(this) { state ->
            handleState(state)
        }
    }

    private fun handleState(state: SearchMoviesState) {
        // Handle loading
        binding.progress.visibility = if (state.isLoading) View.VISIBLE else View.GONE
        
        // Handle movies
        if (state.movies.isNotEmpty()) {
            adapter.resetItems()
            adapter.addItems(state.movies)
            binding.moviesNoData.visibility = View.GONE
        } else {
            adapter.resetItems()
        }
        
        // Handle search hint
        if (state.showSearchHint) {
            binding.moviesNoData.visibility = View.VISIBLE
            binding.moviesNoData.text = getString(R.string.search_movies_hint)
        } else if (state.showNoDataView) {
            binding.moviesNoData.visibility = View.VISIBLE
            binding.moviesNoData.text = getString(R.string.movies_no_data)
        } else {
            binding.moviesNoData.visibility = View.GONE
        }
        
        // Handle loading states
        adapter.setLoading(state.isLoadingNextPage)
        binding.footerProgress.visibility = if (state.isLoadingNextPage) View.VISIBLE else View.GONE
        endlessRecyclerOnScrollListener.setLoading(state.isLoadingNextPage)
        
        // Handle navigation
        state.navigateToMovieDetail?.let { movie ->
            goToMovieDetail(movie)
            viewModel.onNavigationHandled()
        }
        
        // Handle errors
        state.errorMessage?.let { errorMessage ->
            ViewUtils.createSnackbar(binding.toolbar, errorMessage, Snackbar.LENGTH_SHORT).show()
            viewModel.onErrorHandled()
        }
    }

    private fun goToMovieDetail(movie: Movie) {
        selectedMoviePoster?.let { poster ->
            val options = ActivityOptionsCompat.makeSceneTransitionAnimation(
                this,
                poster,
                ViewCompat.getTransitionName(poster) ?: ""
            )
            startActivity(
                MovieDetailActivity.makeIntent(this).putExtra(KEY_MOVIE, movie),
                options.toBundle()
            )
        }
        overridePendingTransition(R.anim.fade_in, R.anim.none)
    }

    override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
        // Menu is empty for this activity
    }

    override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
        return false
    }

    override fun saveInstanceState(outState: Bundle) {
        val currentState = viewModel.state.value
        outState.putString(KEY_SEARCH_QUERY, currentState?.query)
        outState.putParcelable(KEY_PAGINATED_MOVIES, currentState?.paginatedResponse)
    }

    override fun restoreInstanceState(savedState: Bundle) {
        val query = savedState.getString(KEY_SEARCH_QUERY)
        val moviesPaginatedResponse = savedState.getParcelable<PaginatedResponse<Movie>>(KEY_PAGINATED_MOVIES)
        viewModel.processIntent(SearchMoviesIntent.RestoreState(query, moviesPaginatedResponse))
    }


    override fun onItemClick(view: View) {
        val movie = view.tag as? Movie
        movie?.let {
            selectedMoviePoster = view.findViewById(R.id.poster)
            viewModel.processIntent(SearchMoviesIntent.OnMovieItemClick(it))
        }
    }

    companion object {
        const val KEY_SEARCH_QUERY = "SEARCH_QUERY_KEY"
        const val KEY_PAGINATED_MOVIES = "PAGINATED_MOVIES_KEY"

        fun makeIntent(context: Context): Intent = Intent(context, SearchMoviesActivity::class.java)
    }
}